pjlib/src/pjlib-test/string.c

```diff
@@ -289,6 +291,324 @@ static int strcmp_test(void)
+       be 2 (one for the src->ptr content (although the content is a
+       null character), one for the null terminator)
+    */
+    CHECK2_( nulls,  1, 1, 0, "");
+    CHECK2_( "a!",   1, 1, -PJ_ETOOBIG, "");
+    CHECK2_( "a",    1, 1, -PJ_ETOOBIG, "");
+    CHECK2_( "",     0, 1, 0, "");
+    CHECK2_( NULL,   0, 1, 0, "");
+
+    CHECK2_( "abc",  3, 2, -PJ_ETOOBIG, "a");
+    CHECK2_( "ab",   2, 2, -PJ_ETOOBIG, "a");
+    CHECK2_( "a!",   1, 2, 1, "a");
+    CHECK2_( nulls,  1, 2, 0, "");
+    CHECK2_( "!",    0, 2, 0, "");
+    CHECK2_( NULL,   0, 2, 0, "");
+
+    CHECK2_( "abc",  3, 3, -PJ_ETOOBIG, "ab");
+    CHECK2_( "ab",   3, 3, 2, "ab");
+    CHECK2_( nulls,  3, 3, 0, "");
+    CHECK2_( "abc",  2, 3, 2, "ab");
+    CHECK2_( "ab",   2, 3, 2, "ab");
+    CHECK2_( "a",    2, 3, 1, "a");
+    CHECK2_( nulls,  2, 3, 0, "");
+    CHECK2_( "a",    1, 3, 1, "a");
+    CHECK2_( "",     1, 3, 0, "");
+    CHECK2_( "",     0, 3, 0, "");
+    CHECK2_( NULL,   0, 3, 0, "");
+
+    CHECK2_( "abcde",5, 4, -PJ_ETOOBIG, "abc");
+    CHECK2_( "abcd", 4, 4, -PJ_ETOOBIG, "abc");
+    CHECK2_( "abc",  4, 4, 3, "abc");
+    CHECK2_( "ab",   4, 4, 2, "ab");
+    CHECK2_( "a",    4, 4, 1, "a");
+    CHECK2_( nulls,  4, 4, 0, "");
+    CHECK2_( "abc",  3, 4, 3, "abc");
+    CHECK2_( "ab",   3, 4, 2, "ab");
+    CHECK2_( "ab",   2, 4, 2, "ab");
+    CHECK2_( "a",    2, 4, 1, "a");
+    CHECK2_( "",     2, 4, 0, "");
+    CHECK2_( nulls,  2, 4, 0, "");
+    CHECK2_( "a",    1, 4, 1, "a");
+    CHECK2_( nulls,  1, 4, 0, "");
+    CHECK2_( "a",    0, 4, 0, "");
+    CHECK2_( "",     0, 4, 0, "");
+    CHECK2_( NULL,   0, 4, 0, "");
+
+#undef CHECK2_
+
+    return 0;
+}
+
+
+static int verify_strxcat(const char *cdst, const char *src, int dst_size,
+                          int exp_ret, const char *exp_dst)
+{
+    char dst[6];
+    char GUARD = '@';
+    int i, ret;
+
+    PJ_ASSERT_RETURN(src && strlen(cdst) <= 4, -730);
+    PJ_ASSERT_RETURN(strlen(cdst) < dst_size ||
+                     (strlen(cdst)==0 && dst_size==0), -731);
+
+    memset(dst, GUARD, sizeof(dst));
+    if (dst_size) {
+        ret = pj_ansi_strxcpy(dst, cdst, dst_size);
+        PJ_ASSERT_RETURN(ret==strlen(cdst), -732);
+    }
+
+    ret = pj_ansi_strxcat(dst, src, dst_size);
+
+    /* verify return value */
+    if (ret != exp_ret) {
+        PJ_LOG(3,("", "  strxcat \"%s\", \"%s\", dst_size=%d: ret %d!=%d",
+                  cdst, src, dst_size, ret, exp_ret));
+        return -734;
+    }
+
+    /* expected dst content */
+    if (exp_dst) {
+        if (strcmp(dst, exp_dst)) {
+            PJ_LOG(3,("", "  strxcat \"%s\", \"%s\", dst_size=%d: "
+                          "dst content mismatch: \"%s\"!=\"%s\"",  
+                          cdst, src, dst_size, dst, exp_dst));
+            return -736;
+        }
+    }
+
+    /* verify not writing past buffer */
+    for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
+        if (dst[i] != GUARD) {
```

Comment:
To suppress build warning
```
../src/pjlib-test/string.c:550:42: warning: comparison of integers of different signs: 'int' and 'unsigned long' [-Wsign-compare]
     for (i=exp_dst?strlen(exp_dst)+1:0; i<sizeof(dst); ++i) {
```
                                         ~^~~~~~~~~~~~